### PR TITLE
Add extension point to override TargetFramework in consuming project

### PR DESCRIPTION
**Pull request by ViktorHofer**
Required for https://github.com/dotnet/runtime/pull/42826

Tested locally and works as expected.

### CODE REVIEW

File: src/Microsoft.DotNet.Build.Tasks.TargetFramework.Sdk/src/build/Microsoft.DotNet.Build.Tasks.TargetFramework.Sdk.targets [hunk 1]
```diff
@@ -68,22 +68,26 @@
...
-     
-    <PropertyGroup>
-      <_OriginalTargetFramework Condition="'$(_OriginalTargetFramework)' == ''">$(TargetFramework)</_OriginalTargetFramework>
-    </PropertyGroup>
 
-    <ChooseBestP2PTargetFrameworkTask TargetFramework="$(_OriginalTargetFramework)"
-                                      ProjectReferencesWithTargetFrameworks="@(_ProjectRefWithTfms)"
+    <ItemGroup>
+      <_ProjectReferenceWithSetTargetFramework>
+        <ConsumingTargetFramework Condition="'%(ProjectReference.Identity)' == '%(Identity)' and '%(ProjectReference.ConsumingTargetFramework)' != ''">%(ProjectReference.ConsumingTargetFramework)</ConsumingTargetFramework>
```
Comment (ericstj) [inline src/Microsoft.DotNet.Build.Tasks.TargetFramework.Sdk/src/build/Microsoft.DotNet.Build.Tasks.TargetFramework.Sdk.targets:76]: Why do you need to intersect here, I would have expected `_ProjectReferenceWithSetTargetFramework`  to preserve all the metadata that was on the original ProjectReference.

Comment (ViktorHofer) [reply src/Microsoft.DotNet.Build.Tasks.TargetFramework.Sdk/src/build/Microsoft.DotNet.Build.Tasks.TargetFramework.Sdk.targets:76]: you might be right. Need to test this out.

Comment (ViktorHofer) [reply src/Microsoft.DotNet.Build.Tasks.TargetFramework.Sdk/src/build/Microsoft.DotNet.Build.Tasks.TargetFramework.Sdk.targets:76]: No, the projects returned from `GetTargetFrameworks` don't have the original metadata on them:

image

Comment (ericstj) [reply src/Microsoft.DotNet.Build.Tasks.TargetFramework.Sdk/src/build/Microsoft.DotNet.Build.Tasks.TargetFramework.Sdk.targets]: Flows for me:
image

Comment (ViktorHofer) [reply src/Microsoft.DotNet.Build.Tasks.TargetFramework.Sdk/src/build/Microsoft.DotNet.Build.Tasks.TargetFramework.Sdk.targets]: Looked at the output from .NETFramework where I didn't set the attribute. Thanks.

File: src/Microsoft.DotNet.Build.Tasks.TargetFramework.Sdk/src/ChooseBestP2PTargetFrameworkTask.cs [hunk 2]
```diff
@@ -2,13 +2,14 @@
...
 using Microsoft.Build.Framework;
+using System;
 using System.IO;
 
 namespace Microsoft.DotNet.Build.Tasks.TargetFramework.Sdk
 {
     public class ChooseBestP2PTargetFrameworkTask : BuildTask
     {
-        [Required]
+        [Required]
```
Comment (ericstj) [inline src/Microsoft.DotNet.Build.Tasks.TargetFramework.Sdk/src/ChooseBestP2PTargetFrameworkTask.cs]: nit:whitespace

File: src/Microsoft.DotNet.Build.Tasks.TargetFramework.Sdk/src/build/Microsoft.DotNet.Build.Tasks.TargetFramework.Sdk.targets [hunk 3]
```diff
@@ -68,22 +68,20 @@
              BuildInParallel="$(BuildInParallel)"
              RemoveProperties="TargetFramework;RuntimeIdentifier"
              SkipNonexistentTargets="true">
-      <Output TaskParameter="TargetOutputs" ItemName="_ProjectRefWithTfms"/>
+      <Output TaskParameter="TargetOutputs" ItemName="_ProjectReferenceWithSetTargetFramework" />
```
Comment (ericstj) [inline src/Microsoft.DotNet.Build.Tasks.TargetFramework.Sdk/src/build/Microsoft.DotNet.Build.Tasks.TargetFramework.Sdk.targets:71]: nit: naming, should this be _ProjectReferenceWithTargetFrameworks?

File: src/Microsoft.DotNet.Build.Tasks.TargetFramework.Sdk/src/build/Microsoft.DotNet.Build.Tasks.TargetFramework.Sdk.targets [hunk 4]
```diff
@@ -68,22 +68,20 @@
...
                                       RuntimeGraph="$(RuntimeGraph)" >
-      <Output TaskParameter="AnnotatedProjectReferencesWithSetTargetFramework" ItemName="_ProjectReferencesWithBestTargetFrameworks" />
+      <Output TaskParameter="AnnotatedProjectReferencesWithSetTargetFramework" ItemName="_ProjectReferenceWithBestTargetFramework" />
     </ChooseBestP2PTargetFrameworkTask>
     
     <ItemGroup>
       <ProjectReference Remove="@(ProjectReference)" />
-      <ProjectReference Include="@(_ProjectReferencesWithBestTargetFrameworks)" SkipGetTargetFrameworkProperties="true"/>
+      <ProjectReference Include="@(_ProjectReferenceWithBestTargetFramework)"
+                        RemoveMetadata="ConsumingTargetFramework;RawTargetFrameworks"
```
Comment (ericstj) [inline src/Microsoft.DotNet.Build.Tasks.TargetFramework.Sdk/src/build/Microsoft.DotNet.Build.Tasks.TargetFramework.Sdk.targets:87]: Why did you need to RemoveMetadata?  Do we use that pattern elsewhere?

**Review by ericstj — APPROVED** [review]

**Review by ericstj — APPROVED** [review]

File: src/Microsoft.DotNet.Build.Tasks.TargetFramework.Sdk/src/build/Microsoft.DotNet.Build.Tasks.TargetFramework.Sdk.targets (hunk 4 above)
Comment (ViktorHofer) [reply src/Microsoft.DotNet.Build.Tasks.TargetFramework.Sdk/src/build/Microsoft.DotNet.Build.Tasks.TargetFramework.Sdk.targets]: The one metadata was stale and the other one doesn't really matter. For the sake of less risk I'm leaving it as is and reverted my change.
